Enable testing against JDK 13 EA builds

jasontedor opened the pull request:
This commit adds JDK 13 to the CI rotation for testing. For now, we will be testing against JDK 13 EA builds.

jasontedor committed:
Enable testing against JDK 13 EA builds

This commit adds JDK 13 to the CI rotation for testing. For now, we will
be testing against JDK 13 EA builds.

elasticmachine commented:
Pinging @elastic/es-core-infra

jasontedor committed:
Add missing newline

alpar-t reviewed (approved):
LGTM 

Just a suggestion: 
It might be worth doing `s/openjdk12/openjdk13` in `.ci/java-versions.properties` and reverting it before merge, to get a CI run with the new version as part of the PR checks. 
I'm fine if you did this locally too. We had these jobs fail for a while with trivial failures for a number of days until everything was sorted out. Granted I do expect this will become progressively better with the accelerated java releases so just a suggestion.

jasontedor commented:
@atorok It was good a couple of weeks ago, yet I just tried on the latest build (14) just published and there are some issues. I will update this PR.

jasontedor committed:
Fix issues since previous BC

jasontedor committed:
Merge remote-tracking branch 'elastic/master' into jdk-13-ci

* elastic/master: (25 commits)
  Rollup ignores time_zone on date histogram (#40844)
  HLRC: fix uri encode bug when url path starts with '/' (#34436)
  Mutes GatewayIndexStateIT.testRecoverBrokenIndexMetadata
  Docs: Pin two IDs in the rest client (#40785)
  Adds version 6.7.2
  [DOCS] Remind users to include @ symbol when applying license file (#40688)
  HLRC: Convert xpack methods to client side objects (#40705)
  Allow ILM to stop if indices have nonexistent policies (#40820)
  Add an OpenID Connect authentication realm (#40674)
  [DOCS] Rewrite query def (#40746)
  [ML] Changes default destination index field mapping and adds scripted_metric agg (#40750)
  Docs: Drop inline callouts from painless book (#40805)
  [ML] refactoring start task a bit, removing unused code (#40798)
  [DOCS] Document index.load_fixed_bitset_filters_eagerly (#40780)
  Make remote cluster resolution stricter (#40419)
  [ML] Add created_by info to usage stats (#40518)
  SQL: [Docs] Small fixes for CURRENT_TIMESTAMP docs (#40792)
  convert modules to use testclusters (#40804)
  Replace javax activation with jakarta activation (#40247)
  Document restrictions on fuzzy matching when using synonyms (#40783)
  ...

jasontedor committed:
Try JDK 13 in CI

jasontedor commented:
@elasticmachine test this please

jasontedor commented:
@atorok It looks like Gradle is trying to run on JDK 13 which it is not ready for yet. Is it possible in our infrastructure to have it be that Gradle runs on JDK 12 but we set `compiler.home` to 13? I suppose we need another variable in the .ci variables and support for respecting that in CI? Otherwise, this change can not go in yet and we can not start testing against JDK 13.

jasontedor commented:
Relates gradle/gradle#8681

atorok commented:
@jasontedor there isn't currently a way to do that and I was really hopping we could keep it that way as I think this adds a lot of confusion to the CI setup as to which version of Java  is used where and why this is necessary - most of the time on the surface it seems superfluous since the compiler and build java will be the same. 

Do we really think we need this much testing of the compiler this early on ?
I'm asking because if can't build with openjdk13 nobody else can :) so it would seem that 
building with openjdk12 and setting  runtime java to openjdk13 would be sufficient for the time being. 
Of course it would be a nice courtesy to the JDK team to do it, but we could do it as soon as the tool chains we use support it without these workarounds, especially that with the increased release cycles these tools will probably be forced to accelerate their releases as well.

jasontedor commented:
JDK 13 will reach ramp down phase already two months from now. I do not want to be in a position where we are at the last minute investigating any issues we might have compiling with JDK 13, solely because we were blocked on upstream when we could have worked around it.

jasontedor committed:
Try on JDK 13

jasontedor committed:
Drop compiling with JDK 13, it is broken right now

jasontedor committed:
Remove oops

jasontedor committed:
Remove extra whitespace

atorok commented:
I think the best path forward is to test with runtime java on the PR like you have now, revert that before merge, and then set up CI to use different JVMs for Gradle and the compiler, but do it only specifically for Java 13 so it's clear that this is a special case we have to cover now instead of something we want to support on the on going bases. 
This will be based on the premise that upstream will get better at it and the problem will go  away. How does that sound ?

atorok commented:
I tested on this branch to make sure that compiling using a different jdk version than running Gradle still works: 
`./gradlew -Dcompiler.java=13 precommit`
```
=======================================
Elasticsearch Build Hamster says Hello!
  Gradle Version        : 5.3
  OS Info               : Linux 5.0.6-arch1-1-ARCH (amd64)
  Compiler JDK Version  : 13 (Oracle Corporation 13-ea [OpenJDK 64-Bit Server VM 13-ea+15])
  Compiler java.home    : /home/alpar/opt/openjdk-13
  Runtime JDK Version   : 13 (Oracle Corporation 13-ea [OpenJDK 64-Bit Server VM 13-ea+15])
  Runtime java.home     : /home/alpar/opt/openjdk-13
  Gradle JDK Version    : 12 (Oracle Corporation 12 [OpenJDK 64-Bit Server VM 12+33])
  Gradle java.home      : /home/alpar/opt/jdk-12
  Random Testing Seed   : CD6E33CF128AD072
=======================================
```

alpar-t committed:
Allow passing in paths as part of compiler.java and runtime.java

With this change we can now pass in a full path as `compiler.java`
system property.
Previus to this change this was a short-hand for passing in a number
that would be substituted for `JAVA_${number}_HOME`.
We know detect if the property is path or not and act accordingly.

alpar-t committed:
Compile PR with openjdk13  (revert before merge)

alpar-t committed:
Merge remote-tracking branch 'origin/master' into jdk-13-ci

alpar-t committed:
Merge remote-tracking branch 'origin/master' into jdk-13-ci

alpar-t committed:
Fix quoting

jasontedor commented:
Thank you @atorok.

jasontedor committed:
Merge remote-tracking branch 'elastic/master' into jdk-13-ci

* elastic/master: (414 commits)
  Add tasks to build Docker build context artifacts (#41819)
  Replace more uses of immutable map builder (#41823)
  Force selection of calendar or fixed intervals in date histo agg (#33727)
  Switch run task to use real distro (#41590)
  Clarify that path_match also considers object fields. (#41658)
  [DOCS] remove 'es.scripting.update.ctx_in_params' system property for 7.0 (#41643)
  Clarify _doc is a permanent part of certain document APIs. (#41727)
  Remove the jdk directory to save space on bwc tests (#41743)
  Fix full text queries test that start with now (#41854)
  Remove `nonApplicationWrite` from `SSLDriver` (#41829)
  SQL: [Docs] Add example for custom bucketing with CASE (#41787)
  Cleanup Bulk Delete Exception Logging (#41693)
  [DOCS] Rewrite `term` query docs for new format (#41498)
  Mute PermissionsIT#testWhen[...]ByILMPolicy (#41858)
  ReadOnlyEngine assertion fix (#41842)
  [ML] addresses preview bug, and adds check to PUT (#41803)
  Fix javadoc in WrapperQueryBuilder
  Testsclusters use seprate configurations per version (#41504)
  Skip explain fetch sub phase when request holds only suggestions (#41739)
  remove unused import
  ...

jasontedor commented:
@elasticmachine run elasticsearch-ci/default-distro

jasontedor commented:
@elasticmachine run elasticsearch-ci/default-distro

jasontedor committed:
Merge branch 'master' into jdk-13-ci

* master: (176 commits)
  Avoid unnecessary persistence of retention leases (#42299)
  [ML][TEST] Fix limits in AutodetectMemoryLimitIT (#42279)
  [ML Data Frame] Persist and restore checkpoint and position (#41942)
  mute failing filerealm hash caching tests (#42304)
  Safer Wait for Snapshot Success in ClusterPrivilegeTests (#40943)
  Remove 7.0.2 (#42282)
  Revert "Remove 7.0.2 (#42282)"
  [DOCS] Copied note on slicing support to Slicing section. Closes 26114 (#40426)
  Remove 7.0.2 (#42282)
  Mute all ml_datafeed_crud rolling upgrade tests
  Move the FIPS configuration back to the build plugin (#41989)
  Remove stray back tick that's messing up table format (#41705)
  Add missing comma in code section (#41678)
  add 7.1.1 and 6.8.1 versions (#42253)
  Use spearate testkit dir for each run (#42013)
  Add experimental and warnings to vector functions (#42205)
  Fix version in tests since #41906 was merged
  Bump version in BWC check after backport
  Prevent in-place downgrades and invalid upgrades (#41731)
  Mute date_histo interval bwc test
  ...

jasontedor committed:
Merge branch 'master' into jdk-13-ci

* master: (224 commits)
  Use an anonymous inner class instead of lambda for UP-TO-DATE support
  Remove MonitoringClient from x-pack (#42770)
  Remove XPackClient from x-pack (#42729)
  address SmokeTestWatcherWithSecurityIT#testSearchInputWithInsufficientPrivileges (#42764)
  [DOCS] Remove unneeded options from `[source,sql]` code blocks (#42759)
  Remove locale-dependent string checking
  Clarify that inner_hits must be used to access nested fields. (#42724)
  Remove CommonTermsQuery and cutoff_frequency param  (#42654)
  [DOCS] Remove unneeded `ifdef::asciidoctor[]` conditionals (#42758)
  [ML] Add Kibana application privilege to data frame admin/user roles (#42757)
  Re-enable token bwc tests (#42726)
  [ML Data Frame] Refactor stop logic (#42644)
  [ML] [Data Frame] nesting group_by fields like other aggs (#42718)
  Return NO_INTERVALS rather than null from empty TokenStream (#42750)
  Add option to ObjectParser to consume unknown fields (#42491)
  Remove leftover transport module docs (#42734)
  Log the status of security on license change (#42488)
  remove v6.5.x and v6.6.x version constants (#42130)
  Remove transport client docs (#42483)
  Fix docs typo in the certutil CSR mode (#42593)
  ...

jasontedor committed:
Fix javadocs

jasontedor commented:
@elasticmachine test this please

jasontedor committed:
Fix Java version parsing

jasontedor commented:
@elasticmachine run elasticsearch-ci/default-distro

jasontedor committed:
Fix version parsing in webhook https tests

This commit fixes the version parsing in the webhook https integration
tests. The issue here is that the parsing was relying on
java.version. However, java.version can contain additional characters
such as -ea for early access builds. See JEP 233:

Name                            Syntax
------------------------------  --------------
java.version                    $VNUM(\-$PRE)?
java.runtime.version            $VSTR
java.vm.version                 $VSTR
java.specification.version      $VNUM
java.vm.specification.version   $VNUM

Instead, we want java.specification.version.

jasontedor committed:
Fix version parsing in various tests

This commit fixes the version parsing in various tests. The issue here is that
the parsing was relying on java.version. However, java.version can contain
additional characters such as -ea for early access builds. See JEP 233:

Name                            Syntax
------------------------------  --------------
java.version                    $VNUM(\-$PRE)?
java.runtime.version            $VSTR
java.vm.version                 $VSTR
java.specification.version      $VNUM
java.vm.specification.version   $VNUM

Instead, we want java.specification.version.

jasontedor commented on .ci/java-versions.properties:
```diff
@@ -5,6 +5,7 @@
-ES_RUNTIME_JAVA=java11
+ES_RUNTIME_JAVA=openjdk13
 GRADLE_TASK=build
+GRADLE_EXTRA_ARGS=-Dcompiler.java=$HOME/.java/openjdk13
```
Note that this is temporary for testing this pull request. It will be removed before merging.

jasontedor commented on .ci/java-versions.properties:
```diff
@@ -5,6 +5,7 @@
 
 ES_BUILD_JAVA=openjdk12
-ES_RUNTIME_JAVA=java11
+ES_RUNTIME_JAVA=openjdk13
```
Note that this is temporary for testing this pull request. It will be removed before merging.

jasontedor commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -95,6 +95,9 @@ private static File findCompilerJavaHome() {
         String compilerJavaProperty = System.getProperty("compiler.java");
 
         if (compilerJavaProperty != null) {
+            if (compilerJavaProperty.contains("/")) {
```
Note that this is temporary for testing this pull request. It will be removed before merging.

jasontedor commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -106,6 +109,9 @@ private static File findRuntimeJavaHome(final File compilerJavaHome) {
         String runtimeJavaProperty = System.getProperty("runtime.java");
 
         if (runtimeJavaProperty != null) {
+            if (runtimeJavaProperty.contains("/")) {
```
Note that this is temporary for testing this pull request. It will be removed before merging.

jasontedor commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into jdk-13-ci

jasontedor commented:
@elasticmachine run elasticsearch-ci/1

jasontedor commented:
@elasticmachine run elasticsearch-ci/default-distro

jasontedor committed:
Revert changes needed for testing
